File: built-in plugins dsp/mb-reverb-room.cpp
/**
 * MB Room Reverb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Natural room ambience
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_ROOM_H
#define MB_REVERB_ROOM_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbRoom : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-room";
    static constexpr const char* PLUGIN_NAME    = "MB Room Reverb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float size = 0.4f;  // range [0, 1]
    float decay = 1.2f;  // range [0.2, 4]
    float mix = 0.25f;  // range [0, 1]
    };

    MbReverbRoom() = default;
    ~MbReverbRoom() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.size = std::clamp(params.size, 0f, 1f);
        params.decay = std::clamp(params.decay, 0.2f, 4f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Room Reverb
        return input;
    }
};

#endif // MB_REVERB_ROOM_H

File: built-in plugins dsp/mb-reverb-room.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbRoomPlugin: PluginDefinition = { id: 'mb-reverb-room', slug: 'mb-reverb-room', name: 'MB Room Reverb', category: 'effect', type: 'reverb', version: '1.0.0', description: 'Natural room ambience', author: 'Max Booster', parameters: [{ id: 'size', name: 'Size', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }, { id: 'decay', name: 'Decay', type: 'float', defaultValue: 1.2, minValue: 0.2, maxValue: 4, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.25, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { size: 0.4, decay: 1.2, mix: 0.25 } };

export default MbReverbRoomPlugin;
